Preserve _numeric_debug_handle throguh deepcopy and re-export

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129287

Summary:
* Added support for preserving it during deepcopy, need to remap the args since _numeric_debug_handle refers
to the nodes in the graph

TODO: need to fully support re-export, currently the metadata for output node is not preserved

Test Plan:
python test/test_quantization.py -k test_deepcopy_preserve_handle
python test/test_quantization.py -k test_copy_preserve_handle

all related tests:
python test/test_quantization.py -k TestGenerateNumericDebugHandle

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Preserve _numeric_debug_handle throguh deepcopy

Summary:
* renamed numeric_debug_handle to _numeric_debug_handle
* Added support for preserving it during deepcopy, need to remap the args since _numeric_debug_handle refers
to the nodes in the graph

Test Plan:
python test/test_quantization.py -k test_deepcopy_preserve_handle
python test/test_quantization.py -k test_copy_preserve_handle

all related tests:
python test/test_quantization.py -k TestNumericDebugHandle

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Preserve _numeric_debug_handle throguh deepcopy"


Summary:
* renamed numeric_debug_handle to _numeric_debug_handle
* Added support for preserving it during deepcopy, need to remap the args since _numeric_debug_handle refers
to the nodes in the graph

Test Plan:
python test/test_quantization.py -k test_deepcopy_preserve_handle
python test/test_quantization.py -k test_copy_preserve_handle

all related tests:
python test/test_quantization.py -k TestNumericDebugHandle

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhxchen17 commented on torch/fx/graph.py:
```diff
@@ -1348,6 +1348,15 @@ def node_copy(self, node: Node, arg_transform: Callable[[Node], 'Argument'] = la
         assert isinstance(kwargs, dict)
         result_node = self.create_node(node.op, node.target, args, kwargs, node.name, node.type)
         result_node.meta = copy.copy(node.meta)
+        NUMERIC_DEBUG_HANDLE_KEY = "_numeric_debug_handle"
+        if NUMERIC_DEBUG_HANDLE_KEY in node.meta:
+            numeric_debug_handle = node.meta[NUMERIC_DEBUG_HANDLE_KEY]
+            for k, v in dict(numeric_debug_handle).items():
+                if k == "output":
+                    result_node.meta[NUMERIC_DEBUG_HANDLE_KEY][k] = v
+                else:
+                    result_node.meta[NUMERIC_DEBUG_HANDLE_KEY][arg_transform(k)] = v
+
```
@jerryzh168 I guess it might be a better idea to make this piece of logic done through a registration, so that we can allow arbitrary metadata to be preserved on a FX graph.

zhxchen17 replied:
We can sort out the details together if you want.

jerryzh168 replied:
yeah sure, we had some discussion on numeric suite today, and we're planning to change this logic to just preserve a simple id for each node for now, so it will be a bit less complicated

jerryzh168 replied:
> We can sort out the details together if you want.

do you need any inputs from me to come up with the design for the registration?

jerryzh168 committed:
Update on "Preserve _numeric_debug_handle throguh deepcopy"


Summary:
* renamed numeric_debug_handle to _numeric_debug_handle
* Added support for preserving it during deepcopy, need to remap the args since _numeric_debug_handle refers
to the nodes in the graph

Test Plan:
python test/test_quantization.py -k test_deepcopy_preserve_handle
python test/test_quantization.py -k test_copy_preserve_handle

all related tests:
python test/test_quantization.py -k TestNumericDebugHandle

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Preserve _numeric_debug_handle throguh deepcopy"


Summary:
* Added support for preserving it during deepcopy, need to remap the args since _numeric_debug_handle refers
to the nodes in the graph

TODO: need to fully support re-export, currently the metadata for output node is not preserved

Test Plan:
python test/test_quantization.py -k test_deepcopy_preserve_handle
python test/test_quantization.py -k test_copy_preserve_handle

all related tests:
python test/test_quantization.py -k TestNumericDebugHandle

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Preserve _numeric_debug_handle throguh deepcopy and re-export"


Summary:
* Added support for preserving it during deepcopy, need to remap the args since _numeric_debug_handle refers
to the nodes in the graph

TODO: need to fully support re-export, currently the metadata for output node is not preserved

Test Plan:
python test/test_quantization.py -k test_deepcopy_preserve_handle
python test/test_quantization.py -k test_copy_preserve_handle

all related tests:
python test/test_quantization.py -k TestGenerateNumericDebugHandle

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Preserve _numeric_debug_handle throguh deepcopy and re-export"


Summary:
* Added support for preserving it during deepcopy, need to remap the args since _numeric_debug_handle refers
to the nodes in the graph

TODO: need to fully support re-export, currently the metadata for output node is not preserved

Test Plan:
python test/test_quantization.py -k test_deepcopy_preserve_handle
python test/test_quantization.py -k test_copy_preserve_handle

all related tests:
python test/test_quantization.py -k TestGenerateNumericDebugHandle

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhxchen17 commented on test/quantization/pt2e/test_generate_numeric_debug_handle.py:
```diff
@@ -80,3 +81,43 @@ def test_quantize_pt2e_preserve_handle(self):
+    def test_copy_preserve_handle(self):
+        m = TestHelperModules.Conv2dThenConv1d()
+        example_inputs = m.example_inputs()
+        m = capture_pre_autograd_graph(m, example_inputs)
```
can we use the new _export_for_training API? the current API here is already unmaintained.

jerryzh168 replied:
changed the call to `torch.export.export`, except for the quantization one, that has some issues with `model.meta`

zhxchen17 commented:
the rest looks good to me

jerryzh168 committed:
Update on "Preserve _numeric_debug_handle throguh deepcopy and re-export"


Summary:
* Added support for preserving it during deepcopy, need to remap the args since _numeric_debug_handle refers
to the nodes in the graph

TODO: need to fully support re-export, currently the metadata for output node is not preserved

Test Plan:
python test/test_quantization.py -k test_deepcopy_preserve_handle
python test/test_quantization.py -k test_copy_preserve_handle

all related tests:
python test/test_quantization.py -k TestGenerateNumericDebugHandle

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
